File: engine/src/piece.rs
use crate::side::Side;

pub type Piece = u8;

  pub const WHITE_PAWN: Piece = 0;
  pub const WHITE_KNIGHT: Piece = 1;
  pub const WHITE_BISHOP: Piece = 2;
  pub const WHITE_ROOK: Piece = 3;
  pub const WHITE_QUEEN: Piece = 4;
  pub const WHITE_KING: Piece = 5;
  pub const BLACK_PAWN: Piece = 8;
  pub const BLACK_KNIGHT: Piece = 9;
  pub const BLACK_BISHOP: Piece = 10;
  pub const BLACK_ROOK: Piece = 11;
  pub const BLACK_QUEEN: Piece = 12;
  pub const BLACK_KING: Piece = 13;
  pub const NONE: Piece = 14;

  pub const PIECES_COUNT: usize = 15;

  pub fn parse_piece(c: char) -> Piece {
    match c {
      'P' => WHITE_PAWN,
      'N' => WHITE_KNIGHT,
      'B' => WHITE_BISHOP,
      'R' => WHITE_ROOK,
      'Q' => WHITE_QUEEN,
      'K' => WHITE_KING,
      'p' => BLACK_PAWN,
      'n' => BLACK_KNIGHT,
      'b' => BLACK_BISHOP,
      'r' => BLACK_ROOK,
      'q' => BLACK_QUEEN,
      'k' => BLACK_KING,
      _ => NONE,
    }
  }

  pub fn to_piece_char(piece: Piece) -> char {
    match piece {
      WHITE_PAWN => 'P',
      WHITE_KNIGHT => 'N',
      WHITE_BISHOP => 'B',
      WHITE_ROOK => 'R',
      WHITE_QUEEN => 'Q',
      WHITE_KING => 'K',
      BLACK_PAWN => 'p',
      BLACK_KNIGHT => 'n',
      BLACK_BISHOP => 'b',
      BLACK_ROOK => 'r',
      BLACK_QUEEN => 'q',
      BLACK_KING => 'k',
      NONE => ' ',
      _ => '?',
    }
  }

#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Debug)]
pub enum PieceType {
  PAWN,
  KNIGHT,
  BISHOP,
  ROOK,
  QUEEN,
  KING,
}

impl PieceType {
  #[inline(always)]
  pub fn index(self) -> usize {
    self as usize
  }

  #[inline(always)]
  pub fn iter(start: Self, end: Self) -> impl Iterator<Item = Self> {
    (start as u8..=end as u8).map(Self::from)
  }

  pub fn code(self) -> String {
    match self {
      PieceType::PAWN => String::from("p"),
      PieceType::KNIGHT => String::from("n"),
      PieceType::BISHOP => String::from("b"),
      PieceType::ROOK => String::from("r"),
      PieceType::QUEEN => String::from("q"),
      PieceType::KING => String::from("k"),
    }
  }

  pub fn from_san_code(code: char) -> Option<PieceType> {
    match code {
      'P' => Some(PieceType::PAWN),
      'N' => Some(PieceType::KNIGHT),
      'B' => Some(PieceType::BISHOP),
      'R' => Some(PieceType::ROOK),
      'Q' => Some(PieceType::QUEEN),
      'K' => Some(PieceType::KING),
      _ => None
    }
  }
}

impl From<u8> for PieceType {
  #[inline(always)]
  fn from(n: u8) -> PieceType {
    unsafe { std::mem::transmute::<u8, Self>(n) }
  }
}

//
//     public static int flip(int piece) { return piece ^ 8; }
//
    pub fn type_of(piece: Piece) -> PieceType {
        PieceType::from(piece & 0b111)
    }

    pub fn side_of(piece: Piece) -> Side { Side::from(((piece as u8) & 0b1000) >> 3) }

    pub fn make_piece(side: Side, piece_type: PieceType) -> Piece {
        return ((side.index() as u8) << 3) + piece_type.index() as u8;
    }

//     public static String getNotation(int piece){
//         return switch (piece) {
//             case WHITE_PAWN -> "P";
//             case WHITE_KNIGHT -> "N";
//             case WHITE_BISHOP -> "B";
//             case WHITE_ROOK -> "R";
//             case WHITE_QUEEN -> "Q";
//             case WHITE_KING -> "K";
//             case BLACK_PAWN -> "p";
//             case BLACK_KNIGHT -> "n";
//             case BLACK_BISHOP -> "b";
//             case BLACK_ROOK -> "r";
//             case BLACK_QUEEN -> "q";
//             case BLACK_KING -> "k";
//             default -> " ";
//         };
//
//     }

File: ts-engine/src/Piece.ts
export class Piece {
  public static readonly WHITE_PAWN = 0;
  public static readonly WHITE_KNIGHT = 1;
  public static readonly WHITE_BISHOP = 2;
  public static readonly WHITE_ROOK = 3;
  public static readonly WHITE_QUEEN = 4;
  public static readonly WHITE_KING = 5;
  public static readonly BLACK_PAWN = 8;
  public static readonly BLACK_KNIGHT = 9;
  public static readonly BLACK_BISHOP = 10;
  public static readonly BLACK_ROOK = 11;
  public static readonly BLACK_QUEEN = 12;
  public static readonly BLACK_KING = 13;
  public static readonly NONE = 14;

  public static readonly PIECES_COUNT = 15;

  public static flip(piece: number): number {
    return piece ^ 8;
  }

  public static typeOf(piece: number): number {
    return piece & 0b111;
  }

  public static sideOf(piece: number): number {
    return (piece & 0b1000) >>> 3;
  }

  public static makePiece(side: number, pieceType: number): number {
    return (side << 3) + pieceType;
  }

  public static getNotation(piece: number): string {
    switch (piece) {
      case Piece.WHITE_PAWN: return "P";
      case Piece.WHITE_KNIGHT: return "N";
      case Piece.WHITE_BISHOP: return "B";
      case Piece.WHITE_ROOK: return "R";
      case Piece.WHITE_QUEEN: return "Q";
      case Piece.WHITE_KING: return "K";
      case Piece.BLACK_PAWN: return "p";
      case Piece.BLACK_KNIGHT: return "n";
      case Piece.BLACK_BISHOP: return "b";
      case Piece.BLACK_ROOK: return "r";
      case Piece.BLACK_QUEEN: return "q";
      case Piece.BLACK_KING: return "k";
      default: return " ";
    }
  }
}
